add-chain/contract_addresses.go

```diff
@@ -0,0 +1,160 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/ethereum-optimism/superchain-registry/superchain"
+)
+
+type AddressData struct {
+	Address string `json:"address"`
+}
+
+var (
+	// Addresses to retrieve from JSON
+	AddressManager                    = "AddressManager"
+	L1CrossDomainMessengerProxy       = "L1CrossDomainMessengerProxy"
+	L1ERC721BridgeProxy               = "L1ERC721BridgeProxy"
+	L1StandardBridgeProxy             = "L1StandardBridgeProxy"
+	L2OutputOracleProxy               = "L2OutputOracleProxy"
+	OptimismMintableERC20FactoryProxy = "OptimismMintableERC20FactoryProxy"
+	SystemConfigProxy                 = "SystemConfigProxy"
+	OptimismPortalProxy               = "OptimismPortalProxy"
+	ProxyAdmin                        = "ProxyAdmin"
+
+	// Addresses to retrieve from chain
+	SuperchainConfig  = "SuperchainConfig"
+	Guardian          = "Guardian"
+	Challenger        = "Challenger"
+	ProxyAdminOwner   = "ProxyAdminOwner"
+	SystemConfigOwner = "SystemConfigOwner"
+)
+
+func readAddressesFromChain(contractAddresses map[string]string, l1RpcUrl string) error {
+	// SuperchainConfig
+	address, err := castCall(contractAddresses[OptimismPortalProxy], "superchainConfig()(address)", l1RpcUrl)
+	if err != nil {
+		contractAddresses[SuperchainConfig] = ""
+	} else {
+		contractAddresses[SuperchainConfig] = address
+	}
+
+	// Guardian
+	address, err = castCall(contractAddresses[SuperchainConfig], "guardian()(address)", l1RpcUrl)
+	if err != nil {
+		address, err = castCall(contractAddresses[OptimismPortalProxy], "guardian()(address)", l1RpcUrl)
```

Comment:
This error handling could be improved, I think. The cast call above could fail for a number of reasons, and probably only some of them indicate the reason we want to try the second cast call. Some comments explaining which versions of the contracts require which calls would help readability and also help us maintain this section of the code going forward. 


I guess we don't have test coverage for both code paths. Do you think that is something we should add?

Reply:
I think the ideal state is that we detect the version of the contracts in use, and pull in versioned Go bindings that map to those contract ABIs, and then make the appropriate calls using the go-ethereum library plus retry logic. Not a blocker for this PR, but curious to know your thoughts.

Reply:
I'm not sure which versions of contracts expect one call over the other. How can I best determine that information? 

This logic mirrors what is in the existing `add-chain.sh` script, with the added benefit that the conditional to trigger the second `castCall` will more reliably get executed (due to more robust error-handling), compared to the following command, which would only execute the first `cast call`: https://github.com/ethereum-optimism/superchain-registry/blob/main/scripts/add-chain.sh#L126

Reply:
Agreed on the ideal state and addressing that in a follow-up PR

Reply:
Yeah what you have is a definite improvement and should close the associated issue. Agreed about improving in followup PRs.

About contract versions, I don't know the answer to your question. Perhaps @mds1 can pitch in?

Reply:
Agreed with this plan to key off the version in use. We need the mappings (from monorepo release tag to contract version) for bytecode verification also, and since we are doing that in a follow up PR, I will make a note to get this done instead of providing it right now—let me know if that works

Reply:
That's perfect. Thanks @mds1